maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java

```diff
@@ -1349,23 +1349,20 @@ private File determinePom(final CommandLine commandLine, final String workingDir
             alternatePomFile = commandLine.getOptionValue(CLIManager.ALTERNATE_POM_FILE);
         }
 
+        File current = baseDirectory;
         if (alternatePomFile != null) {
-            File pom = resolveFile(new File(alternatePomFile), workingDirectory);
-            if (pom.isDirectory()) {
-                if (modelProcessor != null) {
-                    pom = modelProcessor.locatePom(pom);
-                } else {
-                    pom = new File(pom, "pom.xml");
-                }
-            }
+            current = resolveFile(new File(alternatePomFile), workingDirectory);
+        }
 
-            return pom;
-        } else if (modelProcessor != null) {
-            File pom = modelProcessor.locatePom(baseDirectory);
+        File pom;
+        if (current.isDirectory() && modelProcessor != null) {
+            pom = modelProcessor.locatePom(current);
+        } else {
+            pom = current;
+        }
 
-            if (pom.isFile()) {
-                return pom;
-            }
+        if (pom.isFile()) {
```

Comment:
If the pom locates a directory, would that be a bug that should throw an exception rather than returning null?

Reply:
This method is called twice, a first time when discovering extensions, in which case, the `modelLocator` is not set up yet, and second time to create the actual maven request that will be processed.  So while the `modelLocator` should never return a directory, the `determinePom` method may return `null` during extension discovery.  I'll change the code to only call `determinePom` when `modelLocator` is set up correctly.

Reply:
I tried, but the logic is expected downstream, mostly when you run Maven inside a directory with no `pom.xml`.  In such cases, a `null` value is expected.  Also, the `modelLocator` is required to always return a value, but can point to a non existent file, hence the check.